yahp/hparams.py

```diff
@@ -352,21 +358,45 @@ def get_json_schema(cls) -> Dict[str, Any]:
             hparams_type = type_helpers.HparamsType(class_type_hints[f.name])
             # Name is found in registry, set possible values as types in a union type
             if cls.hparams_registry and f.name in cls.hparams_registry and len(cls.hparams_registry[f.name].keys()) > 0:
-                res['properties'][f.name] = get_registry_json_schema(hparams_type, cls.hparams_registry[f.name])
+                res['properties'][f.name] = get_registry_json_schema(hparams_type, cls.hparams_registry[f.name],
+                                                                     _cls_def, allow_recursion)
             else:
-                res['properties'][f.name] = get_type_json_schema(hparams_type)
+                res['properties'][f.name] = get_type_json_schema(hparams_type, _cls_def, allow_recursion)
             res['properties'][f.name]['description'] = f.metadata['doc']
 
+        _cls_def[cls.__qualname__] = res
+
         return res
 
     @classmethod
-    def dump_jsonschema(cls, f: Union[TextIO, str, pathlib.Path]):
-        """Dump the JSONSchema to ``f``."""
+    def get_json_schema(cls: Type[THparams]) -> Dict[str, Any]:
+        """Generates and returns a JSONSchema dictionary."""
+        _cls_def = {}
+        res = cls._get_json_schema(_cls_def, True)
+
+        # Delete top level name
+        del _cls_def[cls.__qualname__]
+        # Add definitions to top level of schema
+        for key, value in _cls_def.items():
+            if '$defs' not in res:
+                res['$defs'] = {}
+            res['$defs'][key] = value
+
+        return res
+
+    @classmethod
+    def dump_jsonschema(cls: Type[THparams], f: Union[TextIO, str, pathlib.Path], **kwargs: Optional[Dict[str, Any]]):
```

Comment:
Python typing requires only the value of the kwarg to be typed -- omit the Dict[str, ...] part.

```suggestion
    def dump_jsonschema(cls: Type[THparams], f: Union[TextIO, str, pathlib.Path], **kwargs: Any):
```